tooling/miniutil/src/fmt/function.rs

```diff
@@ -0,0 +1,224 @@
+use super::*;
+
+// Formats all functions found within the program.
+// All composite types that are used within `prog` will be added to `comptypes` exactly once.
+pub(super) fn fmt_functions(prog: Program, comptypes: &mut Vec<CompType>) -> String {
+    let mut fns: Vec<(FnName, Function)> = prog.functions.iter().collect();
+
+    // Functions are formatted in the order given by their name.
+    fns.sort_by_key(|(FnName(name), _fn)| *name);
+
+    let mut out = String::new();
+    for (fn_name, f) in fns {
+        let start = prog.start == fn_name;
+        out += &fmt_function(fn_name, f, start, comptypes);
+    }
+
+    out
+}
+
+fn fmt_function(
+    fn_name: FnName,
+    f: Function,
+    start: bool,
+    comptypes: &mut Vec<CompType>,
+) -> String {
+    let fn_name = fmt_fn_name(fn_name);
+
+    // Format function arguments
+    let args: Vec<String> = f
+        .args
+        .iter()
+        .map(|(name, _arg_abi)| {
+            let ident = fmt_local_name(name);
+            let ty = fmt_ptype(f.locals.index_at(name), comptypes);
+
+            format!("{ident}: {ty}")
+        })
+        .collect();
+    let args = args.join(", ");
+
+    // Format return type
+    let mut ret_ty = String::from("none");
+    if let Some((ret, _arg_abi)) = f.ret {
+        ret_ty = fmt_ptype(f.locals.index_at(ret), comptypes);
+    }
+
+    // Format function signature
+    let mut out = if start {
+        format!("start fn {fn_name}({args}) -> {ret_ty} {{\n")
+    } else {
+        format!("fn {fn_name}({args}) -> {ret_ty} {{\n")
+    };
+
+    // Format locals
+    let mut locals: Vec<(LocalName, PlaceType)> = f.locals.iter().collect();
+
+    // The locals are formatted in the order of their names.
+    locals.sort_by_key(|(LocalName(name), _place_ty)| *name);
+
+    for (l, pty) in locals {
+        let local = fmt_local_name(l);
+        let ptype = fmt_ptype(pty, comptypes);
+        out += &format!("  let {local}: {ptype};\n");
```

Comment:
It's probably not worth to rearchitect all this, but FYI you could save a lot of string copies here by (a) passing `out: &mut String` to these functions (rather than have them return a fresh `String`), and (b) using
```suggestion
        write!(out, "  let {local}: {ptype};\n");
```